README.md
# SimpleSearchableSelect (SSS)

SimpleSearchableSelect (SSS) is a lightweight JavaScript library that enhances standard HTML `<select>` elements by providing a simple and searchable selection experience. Notably, this library:

- **Requires No Additional CSS:**
    - SimpleSearchableSelect seamlessly integrates with your existing styles, utilizing the same classes as the original `<select>` element. This means you don't need to add extra CSS to achieve a consistent and cohesive look.

- **No Framework Dependencies:**
    - SimpleSearchableSelect is a standalone library and does not rely on any external JavaScript frameworks. It's designed to be lightweight and easy to integrate into your projects without introducing unnecessary dependencies.

- **Utilizes Native `datalist` Functionality:**
    - The library harnesses the native browser `datalist` functionality to display suggestions, ensuring a familiar and standardized user experience across different platforms.

Enjoy the enhanced selection capabilities without the need for extra styling or external frameworks.

Demo:

- Simple demo: [chalda-pnuzig.github.io/SimpleSearchableSelect/demo/](https://chalda-pnuzig.github.io/SimpleSearchableSelect/demo/)
- Bootstrap 5:  [chalda-pnuzig.github.io/SimpleSearchableSelect/demo/bootstrap5.html](https://chalda-pnuzig.github.io/SimpleSearchableSelect/demo/bootstrap5.html)
- UIkit:  [chalda-pnuzig.github.io/SimpleSearchableSelect/demo/uikit.html](https://chalda-pnuzig.github.io/SimpleSearchableSelect/demo/uikit.html)

## Table of Contents

- [Features](#features)
- [Installation](#installation)
- [Getting Started](#getting-started)
- [Usage](#usage)
- [Options](#options)
- API
    - [setValue()](#setvaluevalue-firechange--true)
    - [getValues()](#getvalues)
    - [clearValue()](#clearvaluevalue--false)
    - [resetValue()](#resetvalue)
    - [destroy()](#destroy)

## Features

- Simple and searchable selection experience for HTML `<select>` elements.
- Autocompletion of input based on available options.
- Support for multiple selections.
- Swipe gestures for removing multiple selections.
- Asynchronous data fetching for dynamic option loading.
- Debouncing of input events for improved performance.

## Installation

SimpleSearchableSelect can be installed via NPM npm install

```bash
npm install @chalda/simple-searchable-select
```

## Getting Started

To use SimpleSearchableSelect, include the library in your project and create an instance of the `SimpleSearchableSelect` class, passing the target `<select>` element as an argument.

```javascript
import {SSS} from './SimpleSearchableSelect.min.js';

const targetSelect = document.getElementById('your-select-element');
const sssInstance  = new SSS(targetSelect);
```

## Usage

After creating an instance of SimpleSearchableSelect, the enhanced functionality will be applied to the target `<select>` element. You can customize the behavior by providing options during instantiation.

```javascript
const targetSelect = document.getElementById('your-select-element');
const options      = {
	inputIntervalTimeout : 300,
	multiple             : true,
	// ...other options
};
const sssInstance  = new SimpleSearchableSelect(targetSelect, options);
```

## Options

When creating an instance of SimpleSearchableSelect, you can customize its behavior by providing options during instantiation. Here are the available options:

- **`inputIntervalTimeout`** (default: `200`):
    - Timeout in milliseconds for debouncing input events. Helps improve performance by delaying processing until after the user has stopped typing.

- **`idPrefix`** (default: `'SSS_'`):
    - Prefix for generating unique element IDs. Each instance of SimpleSearchableSelect will have its own set of unique IDs.

- **`insertPosition`** (default: `'beforebegin'`):
    - DOM insertion position for the input element. Determines where the enhanced input element is inserted relative to the original `<select>` element.

- **`multiple`** (default: same as select):
    - Indicates whether multiple selections are allowed. If not specified, the value is derived from the original `<select>` element.

- **`required`** (default: same as select):
    - Indicates whether the input is required. If not specified, the value is derived from the original `<select>` element.

- **`placeholder`** (default: option with empty string):
    - Indicates the placeholder to show. If not specified, the value is derived from the option in the select element with an empty string ('') as its value

- **`swipeOffset`** (default: `50`):
    - Threshold for swipe gestures. Determines the distance a swipe must cover to trigger removal of a multiple selection.

- **`swipeAnimationSpeed`** (default: `200`):
    - Speed of swipe animation in milliseconds. Controls the duration of the animation when removing a multiple selection with a swipe.

- **`promiseData`** (default: `false`):
    - Asynchronous data fetching function or `false` if not used. If a function, it should return a Promise resolving to an object with 'key' and 'value' properties.

- **`selectedStyle`** (default: `false`):
    - allows customization of the style or behavior of a selected option in datalist.
      If `selectedStyle` is set to a function, this function will be invoked when styling a selected option. The function should accept the reference value of the selected option as an argument and return a string representing the customized style or behavior.

## `setValue(value, fireChange = true)`

Sets the value of the SimpleSearchableSelect instance.

### Parameters

- **`value`**: `string | Array<string>` - The value(s) to set. If the instance is in multiple selection mode, you can pass an array of values.

- **`fireChange`**: `boolean` (optional, default: `true`) - Whether to dispatch a 'change' event after setting the value. If set to `false`, the 'change' event will not be triggered.

### Usage Example

```javascript
// Set a single value
sssInstance.setValue('option-value');

// Set multiple values
sssInstance.setValue(['value1', 'value2']);

// Set value without triggering 'change' event
sssInstance.setValue('new-value', false);
```

### Notes

- If the instance is in multiple selection mode, setting a single value will add that value to the existing selection.
- If the fireChange parameter is set to false, the 'change' event will not be dispatched.

This function is useful for programmatically setting the value of the SimpleSearchableSelect instance, providing flexibility in handling both single and multiple selections.

## `getValues()`

Gets the values currently selected in the SimpleSearchableSelect instance.

### Returns

- **`Object`**: An object containing the selected values. The keys represent the selected values, and the values represent the corresponding display labels.

### Usage Example

```javascript
// Retrieve selected values
const selectedValues = sssInstance.getValues();
console.log(selectedValues);
```

### Notes

- The returned object provides a mapping of selected values to their corresponding display labels.

This function is useful for programmatically retrieving the selected values in the SimpleSearchableSelect instance, allowing you to access the current state of the selection.

## `clearValue(value = false)`

Clears the value of the SimpleSearchableSelect instance.

### Parameters

- **`value`**: `string | Array<string> | false` (optional, default: `false`) - The value(s) to clear. If not specified or set to `false`, the function clears all selected values.

### Usage Example

```javascript
// Clear all selected values
sssInstance.clearValue();

// Clear a specific value
sssInstance.clearValue('value-to-clear');

// Clear multiple values
sssInstance.clearValue(['value1', 'value2']);
```

### Notes

- If the instance is in multiple selection mode, clearing a specific value will remove that value from the selection.

This function is useful for programmatically clearing the selected values in the SimpleSearchableSelect instance, providing flexibility to clear all values or specific ones.

## `resetValue()`

Resets the value of the SimpleSearchableSelect instance, restoring it to the original state.

### Usage Example

```javascript
// Reset the value to the original state
sssInstance.resetValue();
```

### Notes

- This function restores the value of the SimpleSearchableSelect instance to its initial state, based on the original options provided during instantiation.
- If the instance is in multiple selection mode, this function removes all added values and restores the selection to its original state.
- If the instance is in single selection mode, this function resets the selected value to its original state.

This function is useful for reverting the SimpleSearchableSelect instance to its initial state, undoing any modifications made during user interactions or programmatically.

## `destroy()`

Destroys the SimpleSearchableSelect instance, removing all enhancements and restoring the original `<select>` element.

### Usage Example

```javascript
// Destroy the SimpleSearchableSelect instance
sssInstance.destroy();
```

### Notes

- This function removes all enhancements made by SimpleSearchableSelect, reverting the target `<select>` element to its original state.
- It deletes any added elements, event listeners, and data associated with the instance.
- If the target `<select>` element had additional attributes, they are restored to their original values.
- If the target `<select>` element had a <label> associated with it, the association is restored.

This function is useful for cleaning up and removing the enhancements made by SimpleSearchableSelect when the functionality is no longer needed.

CHANGELOG.md
Lock Changelog
=================

(all changes without author notice are by [@chalda-pnuzig](https://github.com/chalda-pnuzig))

## 1.2.0 (2023-12-09)

### Features

- Introduces the ability to specify a custom text for selected options in datalist (`selectedStyle`).

## 1.1.0 (2023-12-05)

### Features

- Introduces the ability to specify a custom placeholder through the options, allowing users to set a personalized placeholder for the input element.

### Bug Fixes

- Resolve issue with duplicate values in multiple select

## 1.0.1 (2023-12-01)

- Correct documentation errors.

## 1.0.0 (2023-11-16)

- Initial release.

SimpleSearchableSelect.js
/**!
 * SSS (aka SimpleSearchableSelect aka SmartSelectSystem)
 *
 * @version   v1.2.0
 * @author    Chalda Pnuzig <chalda＠chalda.it>
 * @copyright Chalda Pnuzig 2023
 * {@link     https://github.com/chalda-pnuzig/SimpleSearchableSelect|GitHub}
 * @license   ISC
 */

/**
 * A JavaScript class that enhances standard HTML <select> elements
 * by providing a simple and searchable selection experience.
 * The SimpleSearchableSelect class offers additional features, including autocompletion
 * and dynamic data loading, to improve the user experience when interacting with select elements.
 *
 * @example
 * // Usage:
 * const targetSelect = document.getElementById('your-select-element');
 * const options = {
 *   inputIntervalTimeout: 300,
 *   multiple: true,
 *   // ...other options
 * };
 * const sssInstance = new SSS(targetSelect, options);
 */

export class SSS {

	/**
	 * @type {HTMLInputElement}
	 */
	$input = document.createElement('input');

	/**
	 * @type {HTMLDataListElement}
	 */
	$dataList = document.createElement('datalist');

	/**
	 * @type {HTMLSelectElement}
	 */
	$target;

	/**
	 * @type {HTMLLabelElement}
	 */
	$label;

	/**
	 * @name PromiseCallback
	 * @type function
	 * @param string
	 * @returns {Promise<Object.<string, string>>}
	 *
	 * it should return a Promise resolving to an object with 'value' and 'text' properties.
	 */

	/**
	 * @typedef {'beforebegin'|'afterbegin'|'beforeend'|'afterend'} DOMInsertion
	 */

	/**
	 * @typedef {function} SelectedStyleFn
	 * @param {string} referenceValue - The  text associated with the selected option.
	 * @returns {string} - The customized text for the selected option.
	 */

	/**
	 * Configuration options for the SSS (Smart Select System) instance.
	 *
	 * @typedef {Object} SSSOptions
	 * @property {number}                [inputIntervalTimeout=200]     - Timeout in milliseconds for debouncing input events.
	 * @property {string}                [idPrefix='SSS_']              - Prefix for generating unique element IDs.
	 * @property {DOMInsertion}          [insertPosition='beforebegin'] - DOM insertion position for the input element.
	 * @property {number}                [swipeOffset=50]               - Threshold for swipe gestures.
	 * @property {number}                [swipeAnimationSpeed=200]      - Speed of swipe animation in milliseconds.
	 * @property {PromiseCallback|false} [promiseData=false]            - Asynchronous data fetching function or false if not used.
	 * @property {false|SelectedStyleFn} [selectedStyle=false]          - If a function is provided, it will be invoked with the reference value of the selected option. If set to `false`, selected options will be disabled.
	 * @property {boolean}               [multiple=undefined]           - Indicates whether multiple selections are allowed. Defaults to the value of the select element.
	 * @property {boolean}               [required=undefined]           - Indicates whether the input is required. Defaults to the value of the select element.
	 * @property {string}                [placeholder=undefined]        - Indicates the placeholder to show. Defaults to the value is derived from the option in the select element with an empty string ('') as its value
	 **/
	options = {
		inputIntervalTimeout : 200,
		idPrefix             : 'SSS_',
		insertPosition       : 'beforebegin',
		swipeOffset          : 50,
		swipeAnimationSpeed  : 200,
		promiseData          : false,
		selectedStyle        : false,
		multiple             : undefined, // Note: multiple and required are intentionally left undefined here
		required             : undefined, // to indicate they will use the values from the input element,
		placeholder          : undefined, // and placeholder is taken from the option in the select with an empty string ('') if present.
	};

	/**
	 * @type {Object.<string, string>}
	 * @private
	 */
	#values = {};

	/**
	 * @type {Object.<string, string>}
	 * @private
	 */
	#refs = {};

	/**
	 * @type {Object.<string, HTMLInputElement>}
	 * @private
	 */
	#clones = {};

	/**
	 * @type {Object.<string, HTMLOptionElement>}
	 * @private
	 */

	#validValues = {};

	/**
	 * @type {Object.<string, HTMLOptionElement>}
	 * @private
	 */
	#datalistOptions = {};

	/**
	 * Initializes a new instance of the SSS class, associating it with a target element and providing optional configuration options.
	 * If the target element already has an SSS instance, returns the existing instance.
	 *
	 * @param {HTMLSelectElement} target - The target select element to associate with the SSS instance.
	 * @param {SSSOptions} options - Optional configuration options for the SSS instance.
	 * @returns {SSS} - The SSS instance associated with the target element.
	 */

	constructor(target, options = {}) {
		if (!target) throw new Error(`SSS: Target select not found!`)
		if (target.hasOwnProperty('SSS')) return target.SSS;
		this.$target          = target;
		this.$target["SSS"]   = this;
		this.options.multiple = target.multiple;
		this.options.required = target.required;
		this.options          = {
			...this.options,
			...options
		};
		this.#init();
		return this;
	}

	/**
	 * Adds a value to the SSS instance for multiple selections, creating a visual clone.
	 *
	 * @param {string} value - The value to add.
	 * @param {boolean} [fireChange=true] - If true, triggers a 'change' event on the target element.
	 * @returns {void}
	 */
	#addMultipleValue(value, fireChange = true) {
		// If a clone for the value already exists, focus on it and reset the input value
		if (typeof this.#clones[value] !== 'undefined') {
			this.#clones[value].focus();
			this.$input.value = '';
			return;
		}

		// If the value is falsy, exit early
		if (!value) return;

		// Set the value reference
		this.#values[value] = this.#refs[value];

		// Create a clone of the input element
		const clone     = this.$input.cloneNode(true);
		let touchStartX = 0;
		let touchEndX   = 0;

		// Event listener for touchstart to capture starting touch position
		clone.addEventListener('touchstart', e => touchStartX = e.changedTouches[0].screenX);

		// Styling and class for the clone
		clone.classList.add('SSS_clone');
		clone.style.setProperty('--SSS-x', '0');
		clone.style.setProperty('--SSS-s', `${this.options.swipeAnimationSpeed}ms`);

		// Set height for the clone if not already set
		clone.style.height ||= window.getComputedStyle(this.$input).height;

		// Custom function to delete the clone
		clone.SSSDelete = () => {
			// Animation and removal logic
			clone.style.setProperty('--SSS-x', '0');
			clone.style.height     = clone.style.padding = clone.style.border = clone.style.fontSize = clone.style.opacity = '0';
			clone.style.transition = 'all var(--SSS-s)';

			setTimeout(() => {
				clone.remove();
				this.#validValues[this.#refs[value]].selected = false;
				delete this.#values[value];
				delete this.#clones[value];

				// Update input's 'required' attribute if needed
				if (this.options.required) {
					this.$input.required = Object.keys(this.#clones).length === 0;
				}

				// Restores the option to an enabled state by setting the 'disabled' property to false.
				this.#datalistOptions[value].disabled = false;

				// resets the option text to its original value.
				this.#datalistOptions[value].text = this.#datalistOptions[value].dataset.original;

				// Trigger 'change' event if specified
				if (fireChange) this.$target.dispatchEvent(new Event('change'));
			}, this.options.swipeAnimationSpeed);
		}

		// Styling for the clone's transform property
		clone.style.transform += ' translateX(var(--SSS-x))';

		// Event listeners for touchend and touchmove to handle swipe deletion
		clone.addEventListener('touchend', e => {
			touchEndX = e.changedTouches[0].screenX;
			if (touchStartX - touchEndX >= this.options.swipeOffset) {
				if (touchEndX < touchStartX) {
					clone.SSSDelete();
				}
			} else {
				clone.style.setProperty('--SSS-x', '0');
			}
		});
		clone.addEventListener('touchmove', e => {
			touchEndX   = e.changedTouches[0].screenX;
			const swipe = touchEndX < touchStartX ? -Math.min(touchStartX - touchEndX, this.options.swipeOffset) : 0;
			clone.style.setProperty('--SSS-x', `${swipe}px`);
		});

		// Store the clone in the #clones object
		this.#clones[value] = clone;

		// Set properties for the clone
		clone.readOnly = true;

		// Event listener for keydown to handle backspace/delete key press
		clone.addEventListener('keydown', (e) => {
			if (['Backspace', 'Delete'].includes(e.key)) {
				clone.SSSDelete();
			}
		});

		// Insert the clone before the input element in the DOM
		this.$input.insertAdjacentElement('beforebegin', clone);

		// Reset the input value and remove 'required' attribute
		this.$input.value    = '';
		this.$input.required = false;

		if (typeof this.options.selectedStyle === 'function') {
			// Apply custom styling to the option text using the selectedStyle function.
			this.#datalistOptions[value].text = this.options.selectedStyle(this.#refs[value]) + '⁣';
		} else {
			// Disable the option if selectedStyle is not a function.
			this.#datalistOptions[value].disabled = true;
		}


		if (this.options.showPicker) {
			try {
				this.$input.showPicker();
			} catch (e) {
			}
		}

		// Update selected status for valid values
		const cloneKeys = Object.keys(this.#clones);
		Object.values(this.#validValues).forEach(o => o.selected = cloneKeys.includes(o.value));

		// Trigger 'change' event if specified
		if (fireChange) this.$target.dispatchEvent(new Event('change'));
	}

	/**
	 * Sets the value of the SSS instance, updating the associated input and triggering a 'change' event if specified.
	 * If the SSS instance is configured for multiple selections, it sets multiple values.
	 *
	 * @param {(string | string[])} value - The value(s) to set.
	 * @param {boolean} [fireChange=true] - If true, triggers a 'change' event on the target element.
	 * @returns {void}
	 */
	setValue(value, fireChange = true) {
		const values = (this.options.multiple ? (Array.isArray(value) ? value : [value]) : [value]);
		values.forEach(v => {
			this.$input.value = this.#refs[v] || '';
			if (this.options.multiple) {
				this.#addMultipleValue(v, false);
			} else {
				this.#values[v]    = this.#refs[v];
				this.$target.value = v;
			}
		});
		if (fireChange) this.$target.dispatchEvent(new Event('change'));
	}

	/**
	 * Clears the value(s) associated with the SSS instance, either for a single or multiple selection.
	 * If the SSS instance is configured for multiple selections, it clears the specified values.
	 * If the SSS instance is configured for a single selection, it clears the value.
	 *
	 * @param {(string | string[] | false)} value - The value(s) to clear. If false, clears all values.
	 * @returns {void}
	 */
	clearValue(value = false) {
		if (this.options.multiple) {
			if (value === false) value = Object.keys(this.#values);
			const valuesToClear = Array.isArray(value) ? value : [value];
			valuesToClear.forEach(v => {
				if (this.#clones[v]) this.#clones[v]["SSSDelete"]();
			});
		} else {
			this.$input.value  = '';
			this.$target.value = '';
		}
	}

	/**
	 * Destroys the SSS instance, restoring the target element to its original state.
	 * If the SSS instance is not attached to the target, the function exits early.
	 * @returns {void}
	 */
	destroy() {
		// Check if the SSS property is not present on the target
		if (!this.$target["SSS"]) return;

		// Restore the visibility of the target element
		this.$target.hidden = false;

		// Remove the reference to this instance from the target object
		delete this.$target["SSS"];

		// Remove associated DOM elements
		this.$dataList.remove();
		this.$input.remove();

		// Restore attributes on the target element
		this.$target.required = this.$target.dataset.sssRequired === '1';
		this.$target.multiple = this.$target.dataset.sssMultiple === '1';
		delete this.$target.dataset.sssRequired;
		delete this.$target.dataset.sssMultiple;

		// Remove cloned elements if the 'multiple' option is enabled
		if (this.options.multiple) Object.values(this.#clones).forEach(el => el.remove());

		// Restore the 'for' attribute of the label, if present
		if (this.$label) {
			this.$label.setAttribute('for', this.$label.dataset.sssOriginal)
			delete this.$label.dataset.sssOriginal;
		}
	}

	/**
	 * Resets the SSS instance, clearing the current value(s) and updating the options based on the target's options.
	 * If the target is a select element with options, it populates the internal data list with valid values.
	 *
	 * @returns {void}
	 */
	resetValue() {
		// Clear current values
		this.clearValue();

		// Array to store selected options
		let selected = [];

		// Iterate over each option in the target element
		this.$target.querySelectorAll('option').forEach(o => {
			let text  = o.textContent;
			let value = o.value;

			// Check if the option is selected and has a value
			if (o.selected && o.value) selected.push(text);

			// Process each option
			if (value) {
				// Create a new Option element and add it to the internal data list
				let newOption = new Option(text);
				this.$dataList.append(newOption);
				newOption.dataset.original   = text;
				this.#datalistOptions[value] = newOption;
				this.#validValues[text]      = o;
				this.#refs[value]            = text;
			} else {
				// Set the input placeholder if the option has no value
				this.$input.placeholder = text;
			}
		});

		// Check if a custom placeholder is specified in the options
		if (this.options.placeholder) {
			this.$input.placeholder = this.options.placeholder;
		}

		// If there are selected options, set the values accordingly
		if (selected.length) {
			if (this.options.multiple) {
				// If multiple selections are allowed, set each selected value
				[...this.$target.selectedOptions].forEach(o => {
					this.setValue(o.value, false);
				});
			} else {
				// If single selection, set the value of the first selected option
				this.$input.value = selected[0];
			}
		}
	}

	/**
	 * Initializes the SSS instance, setting up the necessary elements and event listeners.
	 *
	 * @private
	 * @returns {void}
	 */
	#init() {
		// Copy attributes from the target element to the input element
		this.$target.getAttributeNames().forEach(attr => {
			if (attr.match(/^(name|id)$/i)) return;
			this.$input.setAttribute(attr, this.$target.getAttribute(attr));
		});
		this.$input.setAttribute('autocomplete', 'off');
		// Hide the target element and store its original required and multiple attributes
		this.$target.hidden              = true;
		this.$target.dataset.sssRequired = this.$target.required ? '1' : '0';
		this.$target.dataset.sssMultiple = this.$target.multiple ? '1' : '0';
		this.$target.required            = this.options.required;
		this.$target.multiple            = this.options.multiple;

		// Generate a unique idPrefix based on the existing ids
		let counter = 1;
		while (document.getElementById(this.options.idPrefix + counter + '_list')) {
			counter++;
		}
		this.options.idPrefix += counter;

		// Update label attributes if it exists
		if (this.$target.id) {
			this.$label = document.querySelector(`label[for="${this.$target.id}"]`);
			if (this.$label) {
				this.$label.dataset.sssOriginal = this.$target.id;
				this.$label.setAttribute('for', this.options.idPrefix + '_input');
			}
		}

		// Set ids and attributes for the input and datalist elements
		this.$dataList.id    = this.options.idPrefix + '_list';
		this.$input.id       = this.options.idPrefix + '_input';
		this.$input.required = this.options.required;
		this.$input.multiple = this.options.multiple;
		this.$input.setAttribute('list', this.options.idPrefix + '_list');
		this.$input.classList.add('SSS_input');

		// Insert input element and append datalist element to the body
		this.$target.insertAdjacentElement(this.options.insertPosition, this.$input);
		document.body.append(this.$dataList);

		// Initialize values based on the existing options in the target element
		this.resetValue();

		// Event listeners for input handling
		let start = false;
		this.$input.addEventListener('input', () => {

			// Clear any existing timeout to debounce input
			if (start) clearTimeout(start)

			// Get the current input value
			let inputValue = this.$input.value;

			// Controlla se il valore è uno di quelli generati

			let sel = Object.values(this.#datalistOptions).filter(t => t.text === inputValue);
			if (sel.length === 1) {
				inputValue = sel[0].dataset.original
			}


			if (this.options.promiseData) {
				// Handle logic when promiseData is present
				start = setTimeout(() => {
					if (this.#validValues[inputValue]) {
						// If the value is valid, set it
						this.setValue(this.#validValues[inputValue].value);
					} else {
						// Fetch data from promiseData
						this.options.promiseData(inputValue).then((data) => {
							Object.keys(data).forEach(key => {
								let value = data[key];
								if (typeof this.#validValues[value] === 'undefined') {
									// Create and append new option elements
									let option               = new Option(value, key);
									this.#validValues[value] = option;
									this.$target.append(option);
									this.#refs[key]                 = value;
									let dataListOption              = new Option(value);
									dataListOption.dataset.original = value;
									this.#datalistOptions[key]      = dataListOption;
									this.$dataList.append(dataListOption);
								}
							});
						});
					}
				}, this.options.inputIntervalTimeout);

			} else {
				// Handle logic when promiseData is not present
				let regex = new RegExp('^' + inputValue.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') + '$', 'i')
				let match = Object.keys(this.#validValues).find(entry => entry.match(regex));
				if (match) {
					// If there is a match, set the value
					start = setTimeout(() => {
						this.setValue(this.#validValues[match].value);
					}, this.options.inputIntervalTimeout);
				}
			}
		});

		// Handle blur event
		this.$input.addEventListener('blur', () => {
			// Get the current input value
			let inputValue = this.$input.value;

			// Check if the input value is present and not in the validValues
			if (inputValue && !Object.keys(this.#validValues).includes(inputValue)) {
				// If not in validValues, clear the input value
				this.setValue('');
			}
		});

		// Handle focus event
		if (this.options.showPickerOnFocus) {
			this.$input.addEventListener('focus', () => {
				// Show the picker
				try {
					this.$input.showPicker();
				} catch (e) {
				}
			});
		}

		// Handle keydown events (Backspace, Delete, Tab, Enter)
		this.$input.addEventListener('keydown', (e) => {

			switch (e.key) {
				case 'Backspace':
				case 'Delete':
					// Clear the input value on Backspace or Delete key press
					this.setValue('');
					break;
				case 'Tab':
				case 'Enter':
					// Handle Tab or Enter key press
					const inputValue  = this.$input.value;
					const regex       = new RegExp('' + inputValue.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i')
					const matchingKey = Object.keys(this.#validValues).find(key => key.match(regex));
					// If there is a match, set the corresponding value
					if (matchingKey) this.setValue(this.#validValues[matchingKey].value);
					break;
			}
		});
	}

	/**
	 * Retrieves the values associated with the SSS instance.
	 *
	 * @returns {Object<string, string>} - An object containing the values.
	 */
	getValues() {
		return this.#values;
	}
}